README.md
# fluree-hd-keyring


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
(modification: no type change headlines) and this project adheres to
[Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.1.1] - 2020-09-07

### Changed

- Fix Test errors
- Using v1.0.3 of flureejs-tx,flureejs-request,flureejs-query, flureejs-wallet
- Documentation and test improvements

## [0.1.0] - 2020-08-30

- Basic functionality

[0.1.1]: https://github.com/StylusFrost/fluree-hd-keyring/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/StylusFrost/fluree-hd-keyring/releases/v0.1.0


test/index.spec.ts
/* tslint:disable no-invalid-this */
import assert = require('assert')
import HdKeyring from '../src'
import Query from 'flureejs-query'
import Transaction from 'flureejs-tx'
import Request from 'flureejs-request'
import { isValidAuthID } from 'flureejs-utils'

const sampleMnemonic =
  'finish oppose decorate face calm tragic certain desk hour urge dinosaur mango'
const firstAcct = Buffer.from('Tf8REZpquJyoZeFNUzQT1H3JPzkNKLSGxEA')
const secondAcct = Buffer.from('TeyGJQYZ1YdhGhGJJncVTRqAqJtAMxLTAMf')

describe('hd-keyring', function() {
  let keyring: HdKeyring
  beforeEach(() => {
    keyring = new HdKeyring()
  })

  describe('constructor', function() {
    it('constructs', function() {
      keyring = new HdKeyring({
        mnemonic: sampleMnemonic,
        numberOfAccounts: 2,
      })
      return keyring.getAccounts().then(accounts => {
        assert.equal(Buffer.compare(accounts[0], firstAcct) === 0 ? true : false, true)
        assert.equal(Buffer.compare(accounts[1], secondAcct) === 0 ? true : false, true)
      })
    })
  })

  describe('Keyring.type', function() {
    it('is a class property that returns the type string.', function() {
      const type = HdKeyring.type
      assert.equal(typeof type, 'string')
    })
  })

  describe('#type', function() {
    it('returns the correct value', function() {
      const type = keyring.type
      const correct = HdKeyring.type
      assert.equal(type, correct)
    })
  })

  describe('#serialize empty wallets.', function() {
    it('serializes a new mnemonic', function() {
      keyring
        .serialize()
        .then(output => {
          assert.equal(output['numberOfAccounts'], 0)
          assert.equal(output['mnemonic'], undefined)
        })
        .catch(err => {
          throw err
        })
    })
  })

  describe('#deserialize a private key', function() {
    it('serializes what it deserializes', function() {
      keyring
        .deserialize({
          mnemonic: sampleMnemonic,
          numberOfAccounts: 1,
        })
        .then(() => {
          assert.equal(keyring.wallets.length, 1, 'restores two accounts')
          return keyring.addAccounts(1)
        })
        .then(() => {
          return keyring.getAccounts()
        })
        .then(accounts => {
          assert.equal(Buffer.compare(accounts[0], firstAcct) === 0 ? true : false, true)
          assert.equal(Buffer.compare(accounts[1], secondAcct) === 0 ? true : false, true)
          assert.equal(accounts.length, 2)

          return keyring.serialize()
        })
        .then(serialized => {
          assert.equal(serialized.mnemonic, sampleMnemonic)
        })
        .catch(err => {
          throw err
        })
    })
  })

  describe('#addAccounts', function() {
    describe('with no arguments', function() {
      it('creates a single wallet', function() {
        keyring
          .addAccounts()
          .then(() => {
            assert.equal(keyring.wallets.length, 1)
          })
          .catch(err => {
            throw err
          })
      })
    })

    describe('with a numeric argument', function() {
      it('creates that number of wallets', function() {
        keyring
          .addAccounts(3)
          .then(() => {
            assert.equal(keyring.wallets.length, 3)
          })
          .catch(err => {
            throw err
          })
      })
    })
  })

  describe('#signRequest', function() {
    it('returns the expected value', function() {
      const authID = firstAcct
      const requestParams = {
        from: '0x' + authID.toString('hex'),
        param: '0x7b7d',
        type: '0x6c6973742d736e617073686f7473',
        host: '0x6c6f63616c686f7374',
        db: '0x61756469747a6f6e652d746573746e65742f7365616c6462',
        formattedDate: '0x5475652c2031382041756720323032302030393a30323a353420474d54',
        auth: '0x',
      }
      let request: Request
      return keyring
        .deserialize({
          mnemonic: sampleMnemonic,
          numberOfAccounts: 1,
        })
        .then(() => {
          request = new Request(requestParams)
          return keyring.signRequest(authID, request)
        })
        .then(() => {
          assert.equal(
            request.r.toString('hex'),
            'cbd32e463567fefc2f120425b0224d9d263008911653f50e83953f47cfbef3bc',
          )
          assert.equal(
            request.s.toString('hex'),
            'c7e4e4c4c84ec09afdde99e51dc34f197a14f65e45f95474c45be93bf2552997',
          )
        })
        .catch(reason => {
          assert.ok(false, reason)
        })
    })
  })

  describe('#signTransaction', function() {
    it('returns the expected value', function() {
      const authID = firstAcct
      const txParams = {
        from: '0x' + authID.toString('hex'),
        type: '0x7478',
        db: '0x61756469747a6f6e652d746573746e65742f7365616c6462',
        tx:
          '0x5b7b225f6964223a225f636f6c6c656374696f6e247365616c222c226e616d65223a227365616c222c22646f63223a224120636f6c6c656374696f6e20746f20686f6c642074686520696e666f726d6174696f6e206f6620746865207365616c73227d5d',
        auth: '0x5466386f764864676e445a58724d7a71454c706131787331636664684a696533507761',
        fuel: '0x0f4240',
        nonce: '0x2e',
        expire: '0x017410ddd6e2',
      }
      let tx: Transaction
      return keyring
        .deserialize({
          mnemonic: sampleMnemonic,
          numberOfAccounts: 1,
        })
        .then(() => {
          tx = new Transaction(txParams)
          return keyring.signTransaction(authID, tx)
        })
        .then(() => {
          assert.equal(
            tx.r.toString('hex'),
            'cbd32e463567fefc2f120425b0224d9d263008911653f50e83953f47cfbef3bc',
          )
          assert.equal(
            tx.s.toString('hex'),
            '5f220fb3b29361438f30f5a560641fb91d5193dea39d1f91586ea3194749ae65',
          )
        })
        .catch(reason => {
          assert.ok(false, reason)
        })
    })
  })

  describe('#signQuery', function() {
    it('returns the expected value', function() {
      const authID = firstAcct
      const queryParams = {
        from: '0x' + authID.toString('hex'),
        param: '0x6869207468657265',
        type: '0x7175657279',
        host: '0x6c6f63616c686f7374',
        db: '0x61756469747a6f6e652d746573746e65742f7365616c6462',
        formattedDate: '0x5475652c2031382041756720323032302030393a30323a353420474d54',
        auth: '0x',
      }
      let query: Query
      return keyring
        .deserialize({
          mnemonic: sampleMnemonic,
          numberOfAccounts: 1,
        })
        .then(() => {
          query = new Query(queryParams)
          return keyring.signQuery(authID, query)
        })
        .then(() => {
          assert.equal(
            query.r.toString('hex'),
            'cbd32e463567fefc2f120425b0224d9d263008911653f50e83953f47cfbef3bc',
          )
          assert.equal(
            query.s.toString('hex'),
            'd8fa7b3ecd547cca24cdbc5a9fae3356414accf8081d29263b739bb943d5a245',
          )
        })
        .catch(reason => {
          assert.ok(false, reason)
        })
    })
  })

  describe('custom hd paths', function() {
    it('can deserialize with an hdPath param and generate the same accounts.', function() {
      const hdPathString = `m/44'/60'/0'/0`
      const sampleMnemonic =
        'finish oppose decorate face calm tragic certain desk hour urge dinosaur mango'

      return keyring
        .deserialize({
          mnemonic: sampleMnemonic,
          numberOfAccounts: 1,
          hdPath: hdPathString,
        })
        .then(() => {
          return keyring.getAccounts()
        })
        .then(authIDs => {
          assert.equal(Buffer.compare(authIDs[0], firstAcct) === 0 ? true : false, true)
          return keyring.serialize()
        })
        .then(serialized => {
          assert.equal(serialized.hdPath, hdPathString)
        })
        .catch(reason => {
          assert.ok(false, reason)
        })
    })

    it('can deserialize with an hdPath param and generate different accounts.', function() {
      const hdPathString = `m/44'/60'/0'/1`
      const sampleMnemonic =
        'finish oppose decorate face calm tragic certain desk hour urge dinosaur mango'

      return keyring
        .deserialize({
          mnemonic: sampleMnemonic,
          numberOfAccounts: 1,
          hdPath: hdPathString,
        })
        .then(() => {
          return keyring.getAccounts()
        })
        .then(authIDs => {
          assert.notEqual(Buffer.compare(authIDs[0], firstAcct) === 0 ? true : false, true)
          return keyring.serialize()
        })
        .then(serialized => {
          assert.equal(serialized.hdPath, hdPathString)
        })
        .catch(reason => {
          assert.ok(false, reason)
        })
    })

    describe('create and restore 100 accounts', function() {
      it('should restore same accounts with no problem', async function() {
        // tslint:disable-next-line
        this.timeout(40000)
        for (let i = 0; i < 100; i++) {
          keyring = new HdKeyring({
            numberOfAccounts: 1,
          })
          const originalAccounts = await keyring.getAccounts()
          const serialized = await keyring.serialize()
          const mnemonic = serialized.mnemonic

          keyring = new HdKeyring({
            numberOfAccounts: 1,
            mnemonic,
          })
          const restoredAccounts = await keyring.getAccounts()

          const first = originalAccounts[0]
          const restored = restoredAccounts[0]
          const msg = `Should restore same account from mnemonic: "${mnemonic}"`
          assert.equal(Buffer.compare(first, restored) === 0 ? true : false, true, msg)
        }
      })
    })

    describe('getAppKeyAuthID', function() {
      it('should return a public authID custom to the provided app key origin', async function() {
        const authID = firstAcct

        keyring = new HdKeyring({
          mnemonic: sampleMnemonic,
          numberOfAccounts: 1,
        })
        const appKeyAuthID = await keyring.getAppKeyAuthID(authID, 'someapp.origin.io')

        assert.notEqual(authID, appKeyAuthID)
        assert(isValidAuthID(appKeyAuthID))

        const accounts = await keyring.getAccounts()
        assert.equal(Buffer.compare(accounts[0], firstAcct) === 0 ? true : false, true)
      })

      it('should return different authIDs when provided different app key origins', async function() {
        keyring = new HdKeyring({
          mnemonic: sampleMnemonic,
          numberOfAccounts: 1,
        })

        const authID = firstAcct

        const appKeyAuthID1 = await keyring.getAppKeyAuthID(authID, 'someapp.origin.io')

        assert(isValidAuthID(appKeyAuthID1))

        const appKeyAuthID2 = await keyring.getAppKeyAuthID(authID, 'anotherapp.origin.io')

        assert(isValidAuthID(appKeyAuthID2))

        assert.notEqual(Buffer.compare(appKeyAuthID1, appKeyAuthID2) === 0 ? true : false, true)
      })

      it('should return the same authid when called multiple times with the same params', async function() {
        keyring = new HdKeyring({
          mnemonic: sampleMnemonic,
          numberOfAccounts: 1,
        })

        const authID = firstAcct

        const appKeyAuthID1 = await keyring.getAppKeyAuthID(authID, 'someapp.origin.io')

        assert(isValidAuthID(appKeyAuthID1))

        const appKeyAuthID2 = await keyring.getAppKeyAuthID(authID, 'someapp.origin.io')

        assert(isValidAuthID(appKeyAuthID2))

        assert.equal(Buffer.compare(appKeyAuthID1, appKeyAuthID2) === 0 ? true : false, true)
      })
    })

    describe('signing methods withAppKeyOrigin option', function() {
      it('should signRequest with the expected key when passed a withAppKeyOrigin', function() {
        const authID = firstAcct
        const requestParams = {
          from: '0x' + authID.toString('hex'),
          param: '0x7b7d',
          type: '0x6c6973742d736e617073686f7473',
          host: '0x6c6f63616c686f7374',
          db: '0x61756469747a6f6e652d746573746e65742f7365616c6462',
          formattedDate: '0x5475652c2031382041756720323032302030393a30323a353420474d54',
          auth: '0x',
        }
        let request: Request
        return keyring
          .deserialize({
            mnemonic: sampleMnemonic,
            numberOfAccounts: 1,
          })
          .then(() => {
            request = new Request(requestParams)
            return keyring.signRequest(authID, request, {
              withAppKeyOrigin: 'someapp.origin.io',
            })
          })
          .then(() => {
            assert.equal(
              request.r.toString('hex'),
              'cbd32e463567fefc2f120425b0224d9d263008911653f50e83953f47cfbef3bc',
            )
            assert.equal(
              request.s.toString('hex'),
              'af28a86a68c28b16fe531d3c163ae6d625f193d73aa93629d3b79d7a87d147d2',
            )
          })
          .catch(reason => {
            assert.ok(false, reason)
          })
      })
      it('should signQuery with the expected key when passed a withAppKeyOrigin', function() {
        const authID = firstAcct
        const queryParams = {
          from: '0x' + authID.toString('hex'),
          param: '0x6869207468657265',
          type: '0x7175657279',
          host: '0x6c6f63616c686f7374',
          db: '0x61756469747a6f6e652d746573746e65742f7365616c6462',
          formattedDate: '0x5475652c2031382041756720323032302030393a30323a353420474d54',
          auth: '0x',
        }
        let query: Query
        return keyring
          .deserialize({
            mnemonic: sampleMnemonic,
            numberOfAccounts: 1,
          })
          .then(() => {
            query = new Query(queryParams)
            return keyring.signQuery(authID, query, {
              withAppKeyOrigin: 'someapp.origin.io',
            })
          })
          .then(() => {
            assert.equal(
              query.r.toString('hex'),
              'cbd32e463567fefc2f120425b0224d9d263008911653f50e83953f47cfbef3bc',
            )
            assert.equal(
              query.s.toString('hex'),
              'c03e3ee46dc8474625423fb19825cb12ed276a70fccd0adb4acf4ff7d951c080',
            )
          })
          .catch(reason => {
            assert.ok(false, reason)
          })
      })
      it('should signTransaction with the expected key when passed a withAppKeyOrigin', function() {
        const authID = firstAcct
        const txParams = {
          from: '0x' + authID.toString('hex'),
          type: '0x7478',
          db: '0x61756469747a6f6e652d746573746e65742f7365616c6462',
          tx:
            '0x5b7b225f6964223a225f636f6c6c656374696f6e247365616c222c226e616d65223a227365616c222c22646f63223a224120636f6c6c656374696f6e20746f20686f6c642074686520696e666f726d6174696f6e206f6620746865207365616c73227d5d',
          auth: '0x5466386f764864676e445a58724d7a71454c706131787331636664684a696533507761',
          fuel: '0x0f4240',
          nonce: '0x2e',
          expire: '0x017410ddd6e2',
        }
        let tx: Transaction
        return keyring
          .deserialize({
            mnemonic: sampleMnemonic,
            numberOfAccounts: 1,
          })
          .then(() => {
            tx = new Transaction(txParams)
            return keyring.signTransaction(authID, tx, {
              withAppKeyOrigin: 'someapp.origin.io',
            })
          })
          .then(() => {
            assert.equal(
              tx.r.toString('hex'),
              'cbd32e463567fefc2f120425b0224d9d263008911653f50e83953f47cfbef3bc',
            )
            assert.equal(
              tx.s.toString('hex'),
              '4665d35953072bbf8fa578fc58dbb775c92e3157984d014667ca5757dcc5cca0',
            )
          })
          .catch(reason => {
            assert.ok(false, reason)
          })
      })
    })
  })
})


src/index.ts
import SimpleKeyring from 'fluree-simple-keyring'
import { hdkey } from 'flureejs-wallet'
const bip39 = require('bip39')

// HdKeyring implementation

export default class HdKeyring extends SimpleKeyring {
  static readonly hdPathString = `m/44'/60'/0'/0`
  static type = 'HD Key Tree'
  public readonly type: string = HdKeyring.type
  private root!: hdkey
  private hdWallet!: hdkey
  private hdPath!: string
  private mnemonic!: string
  private opts: any = {}
  constructor(opts?: any) {
    super(opts)
    this.deserialize(opts)
      .then()
      .catch(err => {
        throw err
      })
  }

  // static methods

  // private methods
  private initFromMnemonic(mnemonic: string) {
    this.mnemonic = mnemonic
    const seed = bip39.mnemonicToSeed(mnemonic)
    this.hdWallet = hdkey.fromMasterSeed(seed)
    this.root = this.hdWallet.derivePath(this.hdPath)
  }

  // public instance methods

  /**
   * Serialize all opts
   */
  public serialize() {
    return Promise.resolve({
      mnemonic: this.mnemonic,
      numberOfAccounts: this.wallets.length,
      hdPath: this.hdPath,
    })
  }

  /**
   * Deserialize all opts
   */
  public deserialize(opts: any = {}) {
    this.opts = opts || {}
    this.wallets = []
    this.hdPath = opts['hdPath'] || HdKeyring.hdPathString

    if (this.opts['mnemonic']) {
      this.initFromMnemonic(this.opts['mnemonic'])
    }

    if (this.opts['numberOfAccounts']) {
      return this.addAccounts(this.opts['numberOfAccounts'])
    }

    return Promise.resolve([])
  }

  /**
   * Generate n new AuthID
   */
  public addAccounts(numberOfAccounts = 1) {
    if (!this.root) {
      this.initFromMnemonic(bip39.generateMnemonic())
    }

    const oldLen = this.wallets.length
    const newWallets = []
    for (let i = oldLen; i < numberOfAccounts + oldLen; i++) {
      const child = this.root.deriveChild(i)
      const wallet = child.getWallet()
      newWallets.push(wallet)
      this.wallets.push(wallet)
    }
    return Promise.resolve(newWallets.map(w => w.getAuthID()))
  }
}
